Export generator plugin connection improvements

mcilroyc opened the pull request:
- A few minor optimizations
- In debug mode, try to see if there is plugin websocket (briefly) *before* the usual routine of checking/starting generator.  This Allows us to use a remote generator in debug mode.
- Provide "maxConnectionAttempts to NodeDomain (this requires a new version of generator-connection.   PR forthcoming)

**This requires `bower install` to pick up the new generator-connection version specified in bower.json**

iwehrman commented on src/js/actions/export.js:
```diff
@@ -119,6 +120,40 @@ define(function (require, exports) {
+        return descriptor.playObject(generatorLib.getGeneratorStatus())
+            .then(function (status) {
+                return objUtil.getPath(status, "generatorStatus.generatorStatus") === 1;
+            });
```
Might wanna add a second function to this `then` handler to catch potential failures and return false?

mcilroyc replied:
As written, if _getGeneratorStatus() rejects, it will result in giving up on connecting to the service.  I guess your point is that we should swallow these, and try to enable generator anyway.  I can go with that.

iwehrman commented on src/js/actions/export.js:
```diff
@@ -119,6 +120,40 @@ define(function (require, exports) {
+    var _setGeneratorStatus = function (enabled) {
+        return descriptor.playObject(generatorLib.setGeneratorStatus(enabled))
+            .catch(function (e) {
+                throw new Error("Could not enable generator", e);
```
Unfortunately I believe the second argument will be lost. I think you either have to stringify it and append it to the message, or set `.cause = e`;

mcilroyc replied:
I went the ` + e.message` route.  Are you down with that?

iwehrman commented on src/js/actions/export.js:
```diff
@@ -119,6 +120,40 @@ define(function (require, exports) {
+     * @param {boolean} available
+     */
+    var _setServiceAvailable = function (available) {
+        return this.dispatchAsync(events.export.SERVICE_STATUS_CHANGED, { serviceAvailable: !!available });
```
`@return {Promise}`?

iwehrman commented on src/js/actions/export.js:
```diff
@@ -416,40 +454,60 @@ define(function (require, exports) {
+            return _exportService.init()
+                .bind(this)
+                .then(function () {
+                    log.info("Export: Plugin connection established");
```
I think this is too chatty by default.

mcilroyc replied:
changed to debug.  common, lemme have that at least

iwehrman commented on src/js/actions/export.js:
```diff
@@ -416,40 +454,60 @@ define(function (require, exports) {
+        // to see if the plugin is already running (perhaps on a remote generator connection)
+        var preCheck;
+        if (globalUtil.debug) {
+            log.debug("Doing a quick check to see if Export plugin is available");
```
Ditto. I think we should only add log messages like this if we think they will be useful for some engineers. It doesn't seem like this passes that test.

iwehrman replied:
But, sweet feature.

mcilroyc replied:
removed this debug()

iwehrman commented on src/js/actions/export.js:
```diff
@@ -416,40 +454,60 @@ define(function (require, exports) {
                     .bind(this)
-                    .then(function () {
-                        return this.dispatchAsync(events.export.SERVICE_STATUS_CHANGED, { serviceAvailable: true });
+                    .then(function (enabled) {
```
It would be a little nicer to just return the above promise and move the rest of this chain up to the outer promise?

mcilroyc replied:
derp, yes.

iwehrman commented on src/js/jsx/sections/export/ExportAllPanel.jsx:
```diff
@@ -161,7 +161,7 @@ define(function (require, exports, module) {
                 var exportActions = this.getFlux().actions.export,
                     document = this.state.document;
-                return exportActions.setAllAssetsRequested(document).then(function () {
+                exportActions.setAllAssetsRequested(document).then(function () {
```
Why wait for the next render (is that when these async callbacks happen?) to start this process?

mcilroyc replied:
There seems not to be a reason.  flattened in next commit.  nice catch

iwehrman commented:
Done with a first pass. Back to @mcilroyc.

mcilroyc commented:
back to @iwehrman for a second pass.

mcilroyc commented:
Realized mistake, Taking back

mcilroyc commented:
Ok, re-back to you.  I had hastily refactored the "preCheck" logic incorrectly.  Fixed now.  Please check that again to make sure seems more readable than the previous version.  I introduced some extra "chop" but got rid of the nested `.then()` structure.  The challenge is that I do not want to do *anything* further if preCheck returns true, because the connection is already established.

iwehrman commented:
Needs moar rebase.

mcilroyc commented:
rebased

iwehrman commented on src/js/actions/export.js:
```diff
@@ -119,6 +120,45 @@ define(function (require, exports) {
+            })
+            .catch(function (e) {
+                log.error("Failed to determine generator status: %s", e.message);
+                return Promise.resolve(false);
```
For future reference, you can just return the value here instead of a Promise-boxed value.

iwehrman commented on src/js/actions/export.js:
```diff
@@ -385,30 +428,30 @@ define(function (require, exports) {
-            Promise.reject("Export Service is not available");
+            return _setServiceAvailable.call(this, false)
+                .finally(function () {
+                    return Promise.resolve("Export Service is no longer available");
```
Ditto. Although this is kind of a weird way for the overall promise to resolve.

iwehrman commented on src/js/actions/export.js:
```diff
@@ -416,39 +459,61 @@ define(function (require, exports) {
+                .return(true)
+                .catch(function () {
+                    _exportService = null;
+                    return Promise.resolve(false);
```
Ditto.

iwehrman commented:
Minor nits to be fixed in the next round. Geronimo!
